code/modules/reagents/reagent_containers/iv_bag.dm

```diff
@@ -0,0 +1,180 @@
+			reagents.reaction(injection_target, INGEST, fraction) 						//React the amount we're transfering.
+			reagents.trans_to(injection_target, amount_per_transfer_from_this)
+			update_icon()
+	else		// Drawing
+		if(reagents.total_volume < reagents.maximum_volume)
+			injection_target.transfer_blood_to(src, amount_per_transfer_from_this)
+			for(var/datum/reagent/x in injection_target.reagents.reagent_list) // Pull small amounts of reagents from the person while drawing blood
+				injection_target.reagents.trans_to(src, amount_per_transfer_from_this/10)
+			update_icon()
+
+/obj/item/reagent_containers/iv_bag/attack(mob/living/M, mob/living/user, def_zone)
+	return
+
+/obj/item/reagent_containers/iv_bag/afterattack(atom/target, mob/user, proximity)
+	if(!proximity)
+		return
+	if(!target.reagents)
+		return
+
+	var/mob/living/L
+	if(isliving(target))
+		L = target
+		if(injection_target) // Removing the needle
+			if(L != injection_target)
+				to_chat(user, "<span class='notice'>[src] is already inserted into [injection_target]'s arm!")
+				return
+			if(L != user)
+				L.visible_message("<span class='danger'>[user] is trying to remove [src]'s needle from [L]'s arm!</span>", \
+								"<span class='userdanger'>[user] is trying to remove [src]'s needle from [L]'s arm!</span>")
+				if(!do_mob(user, L))
+					return
+			L.visible_message("<span class='danger'>[user] removes [src]'s needle from [L]'s arm!</span>", \
+								"<span class='userdanger'>[user] removes [src]'s needle from [L]'s arm!</span>")
+			end_processing()
+		else // Inserting the needle
+			if(!L.can_inject(user, 1))
+				return
+			if(L != user)
+				L.visible_message("<span class='danger'>[user] is trying to insert [src]'s needle into [L]'s arm!</span>", \
+									"<span class='userdanger'>[user] is trying to insert [src]'s needle into [L]'s arm!</span>")
+				if(!do_mob(user, L))
+					return
+			L.visible_message("<span class='danger'>[user] inserts [src]'s needle into [L]'s arm!</span>", \
+									"<span class='userdanger'>[user] inserts [src]'s needle into [L]'s arm!</span>")
+			begin_processing(L)
+
+
+/obj/item/reagent_containers/iv_bag/update_icon()
+	overlays.Cut()
+
+	if(reagents.total_volume)
+		var/percent = round((reagents.total_volume / volume) * 10) // We round the 1's place off of our percent for easy image processing.
+		var/image/filling = image('icons/goonstation/objects/iv.dmi', src, "[icon_state][percent]")
+
+		filling.icon += mix_color_from_reagents(reagents.reagent_list)
+		overlays += filling
+	if(ismob(loc))
+		switch(mode)
+			if(IV_DRAW)
+				overlays += "draw"
+			if(IV_INJECT)
+				overlays += "inject"
+
+/obj/item/reagent_containers/iv_bag/proc/set_label(new_label)
+	name = "[initial(name)] ([new_label])"
+
+/obj/item/reagent_containers/iv_bag/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/pen) || istype(I, /obj/item/flashlight/pen))
```

Comment:
one day we should really probably implement a `is_writing_implement` helper. No I don't expect you to do this, just a mental note for myself.